### ISSUE TITLE
complement of roaring bit map does not work correctly

### ISSUE THREAD

**Issue by sirpkt** [druid-io/druid#2175]
Complement of roaring bit map returns incorrect result.
It seems that complement does not work on the last bit of the roaring bitmap.

I used following test code and all tests passed with concise bitmap but failed with roaring bitmap.
Result shows the last bit of roaring bitmap is unchanged even after complement operation.
```@RunWith(Parameterized.class)
public class RoaringBitmapTest
{
  @Parameterized.Parameters
  public static Iterable<Object[]> constructorFeeder()
  {
    return ImmutableList.of(
        new Object[]{new ConciseBitmapFactory()},
        new Object[]{new RoaringBitmapFactory()}
    );
  }

  private final BitmapFactory testBitmapFactory;
  public RoaringBitmapTest(BitmapFactory bitmapFactory)
  {

    this.testBitmapFactory = bitmapFactory;
  }

  @Test
  public void testComplement1()
  {
    final MutableBitmap mutableBitmap = testBitmapFactory.makeEmptyMutableBitmap();
    mutableBitmap.add(0);
    mutableBitmap.add(2);
    mutableBitmap.add(7);
    ImmutableBitmap testBitmap = testBitmapFactory.makeImmutableBitmap(mutableBitmap);
    ImmutableBitmap immutableBitmap = testBitmapFactory.complement(testBitmap, 8);
    String result = immutableBitmap.toString().replaceAll("\\s","").split("\\[|\\]|\\{|\\}")[1];
    Assert.assertEquals("1,3,4,5,6", result);
  }

  @Test
  public void testComplement2()
  {
    final MutableBitmap mutableBitmap = testBitmapFactory.makeEmptyMutableBitmap();
    mutableBitmap.add(0);
    mutableBitmap.add(2);
    ImmutableBitmap testBitmap = testBitmapFactory.makeImmutableBitmap(mutableBitmap);
    ImmutableBitmap immutableBitmap = testBitmapFactory.complement(testBitmap, 4);
    String result = immutableBitmap.toString().replaceAll("\\s","").split("\\[|\\]|\\{|\\}")[1];
    Assert.assertEquals("1,3", result);
  }

  @Test
  public void testComplement3()
  {
    final MutableBitmap mutableBitmap = testBitmapFactory.makeEmptyMutableBitmap();
    mutableBitmap.add(4);
    mutableBitmap.add(5);
    ImmutableBitmap testBitmap = testBitmapFactory.makeImmutableBitmap(mutableBitmap);
    ImmutableBitmap immutableBitmap = testBitmapFactory.complement(testBitmap, 10);
    String result = immutableBitmap.toString().replaceAll("\\s","").split("\\[|\\]|\\{|\\}")[1];
    Assert.assertEquals("0,1,2,3,6,7,8,9", result);
  }
}
```

**Comment by sirpkt** [comment]
As NotFilter relies on complement operation of bitmap, 
this bug probably results in incorrect result at not operation.

**Comment by sirpkt** [comment]
I think that it seems to be from duplicated exclusion of complement range end at flip() of MutableRoaringBitmap and not() of MappeableContainer.

If a range (rangeStart, rangeEnd) is given to flip(), it consider rangeStart as inclusive begining of range and rangeEnd as exclusive end of range, which means it only checks range from rangeStart to rangeEnd - 1.
However, not() excludes the end of given range again, which means it only checks range from rangeStart to rangeEnd - 2.
So, last bit of the given roaring bitmap is excluded from complement operation.

I think it shoud be fixed at RoaringBitmap not Druid :)

**Comment by sirpkt** [comment]
I found that current version of RoaringBitmap already fixed the bug.
And github source of com.metamx.bytebuffer-collections is also updated to use new RoaringBitmap (version 0.5.13).
Unfortunately, however, updated com.metamx.bytebuffer-collections is not released yet.
So, current released version bytebuffer-collections.0.2.1 still depends on old version of RoaringBitmap (0.5.4) which has bug on handling complement.

I have no idea about who is responsible for bytebuffer-collection package.
So, @fjy, can you push the release of new version?

**Comment by drcrallen** [comment]
@sirpkt releasing

**Comment by drcrallen** [comment]
@sirpkt would you mind submitting a PR with the unit tests included?

**Comment by sirpkt** [comment]
I found that ImmutableRoaringBitmap.flip() still has bug even with new version.
(MutableRoaringBitmap and RoaringBitmap works fine)
So, I submitted PR to fix the bug to RoaringBitmap.

I think it needs following
- fix RoaringBitmap bug
- release new version of RoaringBitmap with no flip bug
- release new version of bytebuffer-collections that reflect the new version of RoaringBitmap
- update pom.xml of Druid

Then, I'll add unit test codes to test bitmap complement.
